rfcs/NNNN-decorator-roadmap.md

```diff
@@ -0,0 +1,217 @@
+
+### Non-Goals
+- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)
+
+## Motivation
+
+As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them
+
+Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.
+
+## Detailed Design
+
+Detailed design of the new decorators should be covered in another RFC. This RFC focuses on:
+* That standard decorators become the only way to declare reactive properties
+* That breaking changes are made to remove support for dynamically adding reactive properties, since that will not be used by the new decorators
+* A multi-stage plan for making the changes so that developers can incrementally migrate
+
+### Audiences
+
+If we group our developers by how they relate to decorators, we have four distinct audiences:
+
+- JavaScript developers not yet using decorators
+- TypeScript developers using legacy experimental decorators
+- TypeScript developers not yet using decorators because of forward compatibility concerns
+- TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns
+
+This plan needs to address all audiences during the migration.
+
+### End State
+
+After all stages of the implementation plan are complete:
+
+- Standard decorators are the only decorator implementation
+- Decorators are the one way to define reactive properties
+- Remove `static properties`, `static createProperty()` and associated APIs
+- Remove `static addInitializer()` since the standard decorator API provides this
+- Core decorators (`@property`, `@state`) are exported from the main `reactive-element`, `lit-element` and `lit` modules.
+
+### Standard Decorators
+
+Standard decorators have a much different design than TypeScript's experimental decorators. Rather than getting access to the class and making dynamic, imperative changes to the class, standard decorators have no direct access to the class or the class member they're decorating and must instead return an object describing a replacement for the class member. For instance class accessor decorators return an object with get and set methods.
+
+This makes standard decorator usage must more static than experimental decorators. They can't add new class members, and can't change the type of member they decorate, and ReactiveElement's `static createProperty()` API is simply not callable.
```

Comment:
I suspect that TS will eventually add the ability for decorators to change member types, because in practice this is something they can do at runtime

Reply:
Oh sorry, what I meant by "type" here is actually "kind", ie field, accessor, method, or class (and soon to be method parameter). So a decorator cannot change a field into a method, etc.

Reply:
changed to "kind"